app/services/tax_rate_finder.rb

```diff
@@ -0,0 +1,58 @@
+# Finds tax rates on which an adjustment is based on.
+# For example a packaging fee may contain VAT. This service finds the VAT rate
+# for the tax included in the packaging fee.
+class TaxRateFinder
+  # @return [Array<Spree::TaxRate>]
+  def self.tax_rates_of(adjustment)
+    new.tax_rates(
+      adjustment.originator,
+      adjustment.source,
+      adjustment.amount,
+      adjustment.included_tax
+    )
+  end
+
+  # @return [Array<Spree::TaxRate>]
+  def tax_rates(originator, source, amount, included_tax)
+    find_associated_tax_rate(originator, source) ||
+      find_closest_tax_rates_from_included_tax(amount, included_tax)
+  end
+
+  private
+
+  def find_associated_tax_rate(originator, source)
+    case originator
+    when Spree::TaxRate
+      [originator]
+    when EnterpriseFee
+      enterprise_fee_tax_rates(originator, source)
+    end
+  end
+
+  def enterprise_fee_tax_rates(enterprise_fee, source)
+    case source
+    when Spree::LineItem
+      tax_category = line_item_tax_category(enterprise_fee, source)
+      tax_category ? tax_category.tax_rates.match(source.order) : []
+    when Spree::Order
+      enterprise_fee.tax_category ? enterprise_fee.tax_category.tax_rates.match(source) : []
+    end
+  end
+
+  def line_item_tax_category(enterprise_fee, line_item)
+    if enterprise_fee.inherits_tax_category?
+      line_item.product.tax_category
+    else
+      enterprise_fee.tax_category
+    end
+  end
+
+  # shipping fees and adjustments created from the admin panel have
+  # taxes set at creation in the included_tax field without relation
+  # to the corresponding TaxRate, so we look for the closest one
```

Comment:
thanks for the comment but I've read this code 10 times already and I still don't understand what is this concept of using the closest one... maybe an example would help here.

Reply:
It's weird yes, but it seems like from admin shipping fees and adjustments miss an association to tax rates. Shouldn't we fix this instead of having to infer which tax rate it is? it looks really broken to me. I suggest we create an issue.

Reply:
@luisramos0 I extended that comment and added an example. Let me know if you like it.

@sauloperez I agree that we should fix the interface as well. I opened #3278 for that.

Reply:
wow, thanks @mkllnk now I see why I didnt get it, it's broken and complicated :-) thanks for explaining in detail, explaining/understanding is a good first step to improve it :-)

Reply:
Yes, I learned a lot just by writing that comment. I didn't know any of this when starting to work on this issue.